### PR TITLE
[4.0] Big composer update

### PR DESCRIPTION
**Pull request by wilsonge**
### Summary of Changes
Big round of composer updates. Symfony 5. Last few Joomla Packages to framework 2.0. Codeception to v4 (hence the new packages). Various other updates.

### Testing
There's no one singular thing to test - these packages are used literally on every page. Just navigate around try doing as many actions as possible and report back any new bugs.

### Documentation Changes Required
Nope

### CODE REVIEW

**Comment by astridx** [comment]
I tried a test.

1. `git fetch origin pull/28660/head:feature/upgrade-composer`

2. `git checkout feature/upgrade-composer`

3. I deleted the folder `/libraries/vendor` and run `composer i` and `npm  ci`

4. I made a new installation and discovered a custom component and some language packages. That was OK.

5. I created an item in my custom component. That was OK.

6. I created a new article, make it feature and published. Then I changed to the font end. Everything was fine.

7. I run some tests with my `acceptance.yml`. I use Firefox for the tests. 
[acceptance.txt](https://github.com/joomla/joomla-cms/files/4469131/acceptance.txt)


Here I encountered errors.

```
 ./node_modules/.bin/selenium-standalone install

 ./node_modules/.bin/selenium-standalone start
```

`$ libraries/vendor/bin/codecept run acceptance tests/Codeception/acceptance/01-install/ `

The browser opened automatically and it starts entering data. But the installation was not OK. I finished the installation manually.

`$ libraries/vendor/bin/codecept run acceptance tests/Codeception/acceptance/administrator/` 

[admin.txt](https://github.com/joomla/joomla-cms/files/4469098/admin.txt)

`$ libraries/vendor/bin/codecept run acceptance tests/Codeception/api/` 

The browser opened automatically, but now URL was entered.
[api.txt](https://github.com/joomla/joomla-cms/files/4469099/api.txt)


Now I am not sure if I am doing something wrong or if the tests are not up to date.

File: libraries/src/Filter/InputFilter.php [hunk 1]
```diff
@@ -31,6 +31,13 @@ class InputFilter extends BaseInputFilter
 	 */
 	private $stripUSC = 0;
 
+	/**
+	 * A container for InputFilter instances.
+	 *
+	 * @var    InputFilter[]
+	 * @since  __DEPOY_VERSION__
+	 */
+	protected static $instances = array();
```
Comment (Hackwar) [inline libraries/src/Filter/InputFilter.php:40]: Can you add a blank line after this?

Comment (wilsonge) [reply libraries/src/Filter/InputFilter.php:40]: Done

**Comment by wilsonge** [comment]
> The browser opened automatically and it starts entering data. But the installation was not OK. I finished the installation manually.

Seems weird because I can install ok and so can drone. what exactly failed. i mean i'm upgrading codeception a major version so anything is possible but i need more information to work with I'm afraid

**Comment by wilsonge** [comment]
@rdeutz @Hackwar i'll need help with the system tests failing on media manager after this change

**Comment by jwaisner** [comment]
After testing everything looks good.

*Tested Clean install of Joomla!
*Set error reporting to Maximum
*Installed sample data
*Installed multiple extensions
*created articles, categories, fields, menus, modules, users, and usergroups with no issues
*Deleted the above with no issues
*Indexed using smart search followed by a purge and rebuild with no issues

I am not familiar with system tests so I did not complete those. 

System Info:
*OS - Windows 10
*Database - mysql 5.7.26
*PHP 7.4.4

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on b9b49ec544e2a7d8e9c1064d31e545a515dc0b48<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28660">issues.joomla.org/tracker/joomla-cms/28660</a>.</sub>

**Comment by richard67** [comment]
@wilsonge System tests still failing here. Would it help if you update your branch to latest 4.0-dev?

**Comment by wilsonge** [comment]
> @wilsonge System tests still failing here. Would it help if you update your branch to latest 4.0-dev?

It is up to date unfortunately

**Comment by wilsonge** [comment]
@Hackwar going to need some help here i think

**Comment by Hackwar** [comment]
Looking at the tests, I have no idea why it fails. :-) I installed the same branch and it works fine, but most importantly, according to the screenshot it loads the wrong folder in the media manager.

**Comment by Hackwar** [comment]
Why do we have the changes to InputFilter and SafeHtml in this PR? Shouldn't that be a separate PR?

**Comment by Hackwar** [comment]
I have an assumption and would ask you to check this for me. (I don't want to copy your PR and setup all this, etc...)

It seems as if waitForMediaLoaded() in tests/Codeception/_support/Step/Acceptance/Administrator/Media.php waits for NoSuchElementException and the methods used here don't throw that exception anymore. Maybe simply try with Exception?

**Comment by wilsonge** [comment]
> Why do we have the changes to InputFilter and SafeHtml in this PR? Shouldn't that be a separate PR?

It’s a fallout from going from framework v1 to v2. getInstance is defined in the cms but for some reason the instances static was in the framework. We removed it so it needs to be added here. Safe html is because we typehint the constructor of input filter as array so the null being passed in failed

**Comment by astridx** [comment]
> > @wilsonge The browser opened automatically and it starts entering data. But the installation was not OK. I finished the installation manually.
> 
> Seems weird because I can install ok and so can drone. what exactly failed. i mean i'm upgrading codeception a major version so anything is possible but i need more information to work with I'm afraid

I just tested it with chrome. With chrome the installation is fine.

**Comment by astridx** [comment]
I only looked briefly, but maybe it helps: The Media Manager Check works for me if I comment out the two lines `$I->waitForMediaLoaded();` in the file Joomla4/tests/Codeception/acceptance/administrator/components/com_media/MediaListCest.php

```
	/**
	 * Test check all items.
	 *
	 * @param   Media  $I  Acceptance Helper Object
	 *
	 * @since   4.0.0
	 *
	 * @throws Exception
	 */
	public function deleteMultipleFiles(Media $I)
	{
		$testFileName1 = 'test-image-1.png';
		$testFileName2 = 'test-image-1.jpg';
		$testFileItem1 = MediaListPage::item($testFileName1);
		$testFileItem2 = MediaListPage::item($testFileName2);

		$I->wantToTest('that it is possible to delete multiple files.');
		$I->amOnPage(MediaListPage::$url . $this->testDirectory);
		$I->uploadFile('com_media/' . $testFileName1);
		$I->seeAndCloseSystemMessage('Item uploaded.');
		$I->wait(10);
		$I->waitForElement($testFileItem1);

		// We have to clear the file input, otherwise our method of uploading the file via Codeception will upload it twice
		$I->executeJS('document.getElementsByName(\'file\')[0].value = \'\'');
		//$I->waitForMediaLoaded();
		$I->uploadFile('com_media/' . $testFileName2);
		$I->seeAndCloseSystemMessage('Item uploaded.');
		$I->wait(10);
		//$I->waitForMediaLoaded();
		$I->waitForElement($testFileItem2);
		$I->click($testFileItem1);
		$I->clickHoldingShiftkey($testFileItem2);
		$I->click(MediaListPage::$toolbarDeleteButton);
		$I->waitForElement(MediaListPage::$toolbarModalDeleteButton);
		$I->click(MediaListPage::$toolbarModalDeleteButton);
		$I->seeSystemMessage('Item deleted.');
		$I->waitForElementNotVisible($testFileItem1);
		$I->waitForElementNotVisible($testFileItem2);
		$I->dontSeeElement($testFileItem1);
		$I->dontSeeElement($testFileItem2);
	}

```

**Comment by wilsonge** [comment]
That was the hint i needed fixed

**Comment by jwaisner** [comment]
I have tested this item :white_check_mark: successfully on f7023a0c35015d2b620abb81db90772144ee7e03<br /><br />Retested after latest change and everything still works as expected.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/28660">issues.joomla.org/tracker/joomla-cms/28660</a>.</sub>

**Comment by HLeithner** [comment]
Thanks

File: libraries/src/Filter/InputFilter.php [hunk 2]
```diff
@@ -31,6 +31,13 @@ class InputFilter extends BaseInputFilter
 	 */
 	private $stripUSC = 0;
 
+	/**
+	 * A container for InputFilter instances.
+	 *
+	 * @var    InputFilter[]
+	 * @since  __DEPOY_VERSION__
```
Comment (PhilETaylor) [inline libraries/src/Filter/InputFilter.php:38]: grrrrr

**Pull request merged**